tests/NTest_pixbuf.lua

```diff
@@ -0,0 +1,297 @@
+    initBuffer(buffer1,7,8,9,12)
+    initBuffer(buffer2,7,0,8,12)
+    buffer1:shift(1, nil, 2,3)
+    ok(buffer1 == buffer2, "shift middle right")
+
+    initBuffer(buffer1,7,8,9,12)
+    initBuffer(buffer2,7,9,0,12)
+    buffer1:shift(-1, nil, 2,3)
+    ok(buffer1 == buffer2, "shift middle left")
+
+    -- bounds checks, handle gracefully as string:sub does
+    initBuffer(buffer1,7,8,9,12)
+    initBuffer(buffer2,8,9,12,0)
+    buffer1:shift(-1, pixbuf.SHIFT_LOGICAL, 0,5)
+    ok(buffer1 == buffer2, "shift left out of bound")
+
+    initBuffer(buffer1,7,8,9,12)
+    initBuffer(buffer2,0,7,8,9)
+    buffer1:shift(1, pixbuf.SHIFT_LOGICAL, 0,5)
+    ok(buffer1 == buffer2, "shift right out of bound")
+
+end)
+
+N.test('shift LOGICAL issue #2946', function()
+    local buffer1 = pixbuf.newBuffer(4, 4)
+    local buffer2 = pixbuf.newBuffer(4, 4)
+
+    initBuffer(buffer1,7,8,9,12)
+    initBuffer(buffer2,0,0,0,0)
+    buffer1:shift(4)
+    ok(buffer1 == buffer2, "shift all right")
+
+    initBuffer(buffer1,7,8,9,12)
+    initBuffer(buffer2,0,0,0,0)
+    buffer1:shift(-4)
+    ok(buffer1 == buffer2, "shift all left")
+
+    fail(function() buffer1:shift(10) end, "shifting more elements than buffer size")
+    fail(function() buffer1:shift(-6) end, "shifting more elements than buffer size")
+end)
+
+N.test('shift CIRCULAR', function()
+    local buffer1 = pixbuf.newBuffer(4, 4)
+    local buffer2 = pixbuf.newBuffer(4, 4)
+
+    initBuffer(buffer1,7,8,9,12)
+    initBuffer(buffer2,9,12,7,8)
+    buffer1:shift(2, pixbuf.SHIFT_CIRCULAR)
+    ok(buffer1 == buffer2, "shift right")
+
+    initBuffer(buffer1,7,8,9,12)
+    initBuffer(buffer2,9,12,7,8)
+    buffer1:shift(-2, pixbuf.SHIFT_CIRCULAR)
+    ok(buffer1 == buffer2, "shift left")
+
+    initBuffer(buffer1,7,8,9,12)
+    initBuffer(buffer2,7,9,8,12)
+    buffer1:shift(1, pixbuf.SHIFT_CIRCULAR, 2,3)
+    ok(buffer1 == buffer2, "shift middle right")
+
+    initBuffer(buffer1,7,8,9,12)
+    initBuffer(buffer2,7,9,8,12)
+    buffer1:shift(-1, pixbuf.SHIFT_CIRCULAR, 2,3)
+    ok(buffer1 == buffer2, "shift middle left")
+
+    -- bounds checks, handle gracefully as string:sub does
+    initBuffer(buffer1,7,8,9,12)
+    initBuffer(buffer2,8,9,12,7)
+    buffer1:shift(-1, pixbuf.SHIFT_CIRCULAR, 0,5)
+    ok(buffer1 == buffer2, "shift left out of bound")
+
+    initBuffer(buffer1,7,8,9,12)
+    initBuffer(buffer2,12,7,8,9)
+    buffer1:shift(1, pixbuf.SHIFT_CIRCULAR, 0,5)
+    ok(buffer1 == buffer2, "shift right out of bound")
+
+    initBuffer(buffer1,7,8,9,12)
+    initBuffer(buffer2,12,7,8,9)
+    buffer1:shift(1, pixbuf.SHIFT_CIRCULAR, -12,12)
+    ok(buffer1 == buffer2, "shift right way out of bound")
+
+end)
+
+N.test('sub', function()
+    local buffer1 = pixbuf.newBuffer(4, 4)
+    local buffer2 = pixbuf.newBuffer(4, 4)
+    initBuffer(buffer1,7,8,9,12)
+    buffer1 = buffer1:sub(4,3)
+    ok(eq(buffer1:size(), 0), "sub empty")
+
+    buffer1 = pixbuf.newBuffer(4, 4)
+    initBuffer(buffer1,7,8,9,12)
```

Comment:
One line missing here: test fails 
```suggestion
    buffer2 = ws2812.newBuffer(2, 4)
    initBuffer(buffer1,7,8,9,12)
```